README.md
# Deployer action

This action handles deployment via deployer

```
  - name: Deploy
    uses: wearebuilders/action-deployer@master
    env:
      SSH_PRIVATE_KEY: ${{ secrets.PRIVATE_KEY }}
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      SLACK_WEBHOOK_TOKEN: ${{ secrets.SLACK_WEBHOOK_TOKEN }}
```


New way to configure git hub action. Direct docker image

```
name: Auto deployment production/staging

on:
  push:
    branches:
      - 'master'
      - 'staging'
      - 'production'

jobs:
  build:
    name: Deployment
    runs-on: self-hosted

    container:
      image: docker://wimwinterberg/action-deployer:1.0.1-7.3-cli-alpine

    steps:
      - name: Deploy
        run: deploy.sh
        env:
          SSH_PRIVATE_KEY: ${{ secrets.DEPLOYMENT_PRIVATE_KEY }}
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          SLACK_WEBHOOK_TOKEN: ${{ secrets.SLACK_WEBHOOK_TOKEN }}
          COMPOSER_AUTH_JSON: ${{ secrets.COMPOSER_AUTH_JSON }}

```


CHANGELOG.md
v1.1.20
* using php 7.4
* using google dns server

# v1.1.7
* added ScriptHandler SEF logic to deployer
* composer install, cache:clear, migration, init cache
* update deployer to version 6.8.0
 
# v1.0.2
* added shared logic for sef projects

# v1.0.0
* removed node, npm, yarn
* added make for building docker image
* added duration in slack message
* should use pre build docker image

# v0.0.14
* Added node, npm and yarn to docker

# v0.0.13
* Add verbose level

# v0.0.12
* Added custom tasks for sf - crobjob scanning and encore build

# v0.0.11
* Sym-linking the build directory instead of copying

# v0.0.10
* Added shared .env.local file used for building
* Added symlink for private_html (used for https)

# v0.0.9
* Altered build process. Also copying build files instead of symlinking

# v0.0.8
* Added wab-react. Used for apps and react websites 

# v0.0.7
* Added setup permissions
* Added deploy additional info
* Added slack notification when deployment starts
* Display task info when creating github release

# v0.0.6
* Support for WAB SF4 projects
* Moved shared tasks from SEF to bootstrap
* Set write mode to chmod since ACL is enabled default for IPS containers
* Copy vendor directory for SF

# v0.0.5
* Added functionality to create a release
* Added functionality to set composer authentication

# v0.0.4
* Added failure slack message
* Added supported role batch - will not run migration
* Forces annsicon for coloring support in github

# v0.0.3
* download deploy configuration from repo

# v0.0.2
* fixed typo in var
* adjusted slack message

# v0.0.1
* Initial version


test.php
<?php

namespace Deployer;

require_once 'recipes/sef.php';


deploy.sh
#!/bin/sh -l

set -e

eval $(ssh-agent -s)

echo -e "StrictHostKeyChecking no" >> /etc/ssh/ssh_config
echo "$SSH_PRIVATE_KEY" | tr -d '\r' > /tmp/id_rsa
chmod 600 /tmp/id_rsa
ssh-add /tmp/id_rsa

# region $INPUT_VERBOSELEVEL
if [ -z "$INPUT_VERBOSELEVEL" ]; then
  INPUT_VERBOSELEVEL='v'
fi
#endregion

# region $INPUT_DEPLOYERFILELOCATION
if [ -z "$INPUT_DEPLOYERFILELOCATION" ]; then
  INPUT_DEPLOYERFILELOCATION='.deployment/deploy.php'
fi
#endregion

# change DNS
echo "nameserver 8.8.8.8" > /etc/resolv.conf

# download deployer from repository
curl -sS -H 'Authorization: token '"${GITHUB_TOKEN}"'' --location --request GET 'https://raw.githubusercontent.com/'"${GITHUB_REPOSITORY}"'/'"${GITHUB_SHA}"'/'"${INPUT_DEPLOYERFILELOCATION}"'' >deploy.php

# forces coloring - WIP need to find a better solution for this - https://github.com/symfony/console/blob/fb5419f837e0bd960696ebfd143f213dd5c8f744/Output/StreamOutput.php#L100
export TERM_PROGRAM=Hyper

deployer --version
deployer --file=./deploy.php -$INPUT_VERBOSELEVEL deploy


entrypoint.sh
#!/bin/sh -l

set -e

eval $(ssh-agent -s)

echo -e "StrictHostKeyChecking no" >> /etc/ssh/ssh_config
echo "$SSH_PRIVATE_KEY" | tr -d '\r' > /tmp/id_rsa
chmod 600 /tmp/id_rsa
ssh-add /tmp/id_rsa

# region $INPUT_VERBOSELEVEL
if [ -z "$INPUT_VERBOSELEVEL" ]
then
  INPUT_VERBOSELEVEL='v'
fi
#endregion

# region $INPUT_DEPLOYERFILELOCATION
if [ -z "$INPUT_DEPLOYERFILELOCATION" ]
then
  INPUT_DEPLOYERFILELOCATION='.deployment/deploy.php'
fi
#endregion

# download deployer from repository
curl -sS -H 'Authorization: token '"${GITHUB_TOKEN}"'' --location --request GET 'https://raw.githubusercontent.com/'"${GITHUB_REPOSITORY}"'/'"${GITHUB_SHA}"'/'"${INPUT_DEPLOYERFILELOCATION}"'' > deploy.php

# forces coloring - WIP need to find a better solution for this - https://github.com/symfony/console/blob/fb5419f837e0bd960696ebfd143f213dd5c8f744/Output/StreamOutput.php#L100
export TERM_PROGRAM=Hyper

deployer --version
deployer --file=./deploy.php -$INPUT_VERBOSELEVEL deploy


recipes/sef.php
<?php

namespace Deployer;

use Utils\ProjectUtils;

require_once __DIR__ . '/bootstrap_github_action.php';

// ahared files/dirs between deploys
set('shared_dirs', [
    'var/log',
]);

set('shared_files', [
    '.env.local',
]);

// copy dirs
set('copy_dirs', ['vendor']);

// writable dirs by web server
set('writable_dirs', ['var/cache/{{sef_env}}']);
set('writable_mode', 'chmod');
set('writable_chmod_mode', '0777');

set('bin/console', '{{bin/php}} {{release_path}}/bin/console');
set('console_options', function () {
    return '';
});
set('composer_options', '{{composer_action}} --verbose --prefer-dist --no-progress --no-dev --optimize-autoloader'); // removed --no-interaction ~ we like to see more info :)

// region sef app version
set(
    'app_version',
    function () {

        $cmd = "curl -sS -H 'Authorization: token {{github_token}}' --location --request GET 'https://raw.githubusercontent.com/{{repository_name}}/{{revision}}/application/config/application/Config.php'";
        $configContents = runLocally($cmd);

        return (new ProjectUtils())->parseAppVersion(ProjectUtils::APP_TYPE_SEF, $configContents);
    }
);
// endregion

// region sef env
set(
    'sef_env',
    function () {
        $stage = get('stage') ?? get('default_stage');
        switch ($stage) {
            case 'staging':
                $sefEnv = 'staging';
                break;
            case 'production':
                $sefEnv = 'prod';
                break;
            default:
                echo 'Unable to determine sef environment';
                exit(1);
        }

        return $sefEnv;
    }
);

// endregion

// endregion

// region tasks
desc('Running migration');
task(
    'deploy:migrate',
    function () {
        $roles = has('roles') ? get('roles') : null;
        if (!is_array($roles) || !in_array('batch', $roles)) {
            $cmd = 'cd {{release_path}} && {{bin/php}} application/utils/bin/run.php {{sef_env}} migration/migrate';

            $result = run($cmd);
            writeln($result);
        }
    }
);

desc('Display additional info');
task(
    'deploy:info-stage',
    function () {
        writeln('Deployment target is <info>{{stage}}</info>, using SEF environment <info>{{sef_env}}</info>, deploying app version <info>{{app_version}}</info>');
    }
);

desc('Clear cache');
task('deploy:cache:clear', function () {
    run('cd {{release_path}} && rm -rf var/cache/{{sef_env}}/*');
});

desc('Warmup cache');
task('deploy:cache:warmup', function () {
    $result = run('cd {{release_path}} && {{bin/console}} cache:warmup');
    writeln($result);
});

task(
    'deploy',
    [
        'deploy:info',
        'deploy:prepare',
        'deploy:lock',
        'deploy:release',
        'deploy:update_code',
        'composer-set-auth',
        'deploy:shared',
        'deploy:writable',
        'deploy:vendors',
        'deploy:cache:clear',
        'deploy:cache:warmup',
        'deploy:migrate',
        'deploy:symlink',
        'deploy:unlock',
        'cleanup',
        'success',
    ]
);
// endregion


recipes/wab-react.php
<?php

namespace Deployer;

require 'recipes/bootstrap_github_action.php';

set('shared_files', ['.env.local', '.env']);
set('shared_dirs', ['node_modules']); // can be shared, only needed for building
set('copy_dirs', []);
set('writable_dirs', []);
set(
    'app_version',
    function () {

        // @TODO: we can use the package.json to read the version

        $cmd            = "curl -sS -H 'Authorization: token {{github_token}}' --location --request GET 'https://raw.githubusercontent.com/{{repository_name}}/{{revision}}/CHANGELOG.md'";
        $configContents = runLocally($cmd);
        if (preg_match('/^#\s+v?(.*)\s*$/m', $configContents, $matches)) {
            $appVersion = 'v' . $matches[1];
        } else {
            $appVersion = 'vX.Y.Z';
        }

        return $appVersion;
    }
);

desc('Set symlink');
task(
    'create-public-html-symlink',
    function () {
        run("cd {{release_path}} && ln -s build public_html");
    }
);

desc('Build');
task(
    'build',
    function () {
        run('cd {{release_path}} && yarn install --force');
        run('cd {{release_path}} && chmod 755 -R node_modules/.bin/*');
        run('cd {{release_path}} && NODE_ENV=production yarn run build');
    }
);

desc('Display additional info');
task(
    'deploy:info-stage',
    function () {
        writeln('Deployment target is <info>{{stage}}</info>, deploying app version <info>{{app_version}}</info>');
    }
);

task(
    'deploy',
    [
        'deploy:info',
        'deploy:prepare',
        'deploy:lock',
        'deploy:release',
        'download-revision',
        'extract-revision',
        'deploy:shared',
        'build',
        'create-public-html-symlink',
        'create-private-html-symlink',
        'deploy:setup-permissions',
        'deploy:symlink',
        'deploy:unlock',
        'cleanup',
        'success',
    ]
);
// endregion


Utils/ProjectUtils.php
<?php
declare(strict_types=1);

namespace Utils;

use Symfony\Component\Yaml\Exception\ParseException;
use Symfony\Component\Yaml\Yaml;

/**
 * Class ProjectUtils
 */
class ProjectUtils
{
    const APP_TYPE_SF4 = 'SF4';
    const APP_TYPE_SEF = 'SEF';

    /**
     * Parse change log
     *
     * @param string $changelog
     * @return array
     */
    public function parseChangelog(string $changelog): array
    {
        $retValue = [];
        $key      = null;

        foreach (preg_split("/\r\n|\n|\r/", $changelog) as $line) {
            $line = trim($line);
            if (preg_match('/^#\s?v([0-9]*\.[0-9]*\.[0-9]*)$/', $line, $matches)) {
                $key            = 'v' . $matches[1];
                $retValue[$key] = [];
            } elseif (preg_match('/^\* \\\\#([0-9]*)\s(.*)$/', $line, $matches)) {
                $retValue[$key][] = [
                    'number' => $matches[1],
                    'title'  => $matches[2],
                ];
            } elseif (mb_strlen($line) > 0) {
                $retValue[$key][] = ['text' => $line];
            }
        }

        return $retValue;
    }

    /**
     * Parse app version from config
     *
     * @param string $appType
     * @param string $configContent
     * @return string|null
     */
    public function parseAppVersion(string $appType, string $configContent): ?string
    {
        $retValue = null;

        switch ($appType) {
            case self::APP_TYPE_SF4:
                try {
                    $appConfig  = Yaml::parse($configContent);
                    $parameters = $appConfig['parameters'] ?? [];
                    $retValue   = $parameters['app.release'] ?? null;
                } catch (ParseException $exception) {
                }
                break;
            case self::APP_TYPE_SEF:
                if (preg_match('/^\s*self::DEPLOYMENT_VERSION\s*=>\s*\'(.*)\',\s*$/m', $configContent, $matches)) {
                    $retValue = $matches[1];
                }
                break;
        }

        if ($retValue !== null && strtolower(substr($retValue, 0, 1)) != 'v') {
            $retValue = 'v' . $retValue;
        }

        return $retValue;
    }

    /**
     * Change log version to string
     *
     * @param array $items
     * @return string
     */
    public function changeLogVersionToString(array $items): string
    {
        $contentLines = [];
        foreach ($items as $item) {
            if (array_key_exists('text', $item)) {
                $contentLines[] = $item['text'];
            } else {
                $contentLines[] = sprintf('* \#%1$d %2$s', $item['number'], $item['title']);
            }
        }
        $contentLines[] = '';

        return trim(implode(PHP_EOL, $contentLines));
    }

}


recipes/wab-reactV2.php
<?php

namespace Deployer;

require 'recipes/bootstrap_github_action.php';

set('shared_files', ['.env.local', '.env']);
set('shared_dirs', []); // can be shared, only needed for building
set('copy_dirs', []);
set('writable_dirs', []);
set(
    'app_version',
    function () {

        // @TODO: we can use the package.json to read the version

        $cmd            = "curl -sS -H 'Authorization: token {{github_token}}' --location --request GET 'https://raw.githubusercontent.com/{{repository_name}}/{{revision}}/CHANGELOG.md'";
        $configContents = runLocally($cmd);
        if (preg_match('/^#\s+v?(.*)\s*$/m', $configContents, $matches)) {
            $appVersion = 'v' . $matches[1];
        } else {
            $appVersion = 'vX.Y.Z';
        }

        return $appVersion;
    }
);

desc('Set symlink');
task(
    'create-public-html-symlink',
    function () {
        run("cd {{release_path}} && ln -s build public_html");
    }
);



desc('Download revision tar');
task(
    'download-revision',
    function () {

        $repositoryName = get('repository_name', 'xxx');
        $revision       = get('revision', 'xxx');
        $token          = get('github_token', 'xxx');

        runLocally("curl -sS -H 'Authorization: token $token' --location --request GET 'https://api.github.com/repos/$repositoryName/tarball/$revision' > repo.tar.gz");
    }
);

desc('Extract downloaded revision tar');
task(
    'extract-revision',
    function () {
        $repositoryName = get('repository_name', 'xxx');
        $revision       = get('revision', 'xxx');

        $directionName = str_replace('/', '-', $repositoryName) . '-' . $revision . '/';
        runLocally("tar -xvf repo.tar.gz  $directionName");
        runLocally("(mv $directionName{*,.[^.]*,..?*} . 2>/dev/null || true)"); // move all files included dot files (when a project does not have dot files ignore error
        runLocally("rm -rf $directionName");
        runLocally("rm repo.tar.gz");
    }
);

desc('Build');
task(
    'build',
    function () {
        runLocally('ls -slah');
        runLocally('yarn install --force');
        runLocally('NODE_ENV=production yarn run build');
    }
);

desc('Display additional info');
task(
    'deploy:info-stage',
    function () {
        writeln('Deployment target is <info>{{stage}}</info>, deploying app version <info>{{app_version}}</info>');
    }
);

task(
    'deploy',
    [
        'deploy:info',
        'deploy:prepare',
        'deploy:lock',
        'deploy:release',
        'download-revision',
        'extract-revision',
        'deploy:shared',
        'build',
        'create-public-html-symlink',
        'create-private-html-symlink',
        'deploy:setup-permissions',
        'deploy:symlink',
        'deploy:unlock',
        'cleanup',
        'success',
    ]
);
// endregion


recipes/wab-symfony.php
<?php

namespace Deployer;

use Utils\ProjectUtils;

require_once __DIR__ . '/bootstrap_github_action.php';

require 'recipe/symfony4.php';

// region sf tasks
desc('Scanning for cronjob commands');
task(
    'cron-scan',
    function () {
        $result = run('cd {{release_path}} && {{bin/console}} shapecode:cron:scan');
        writeln($result);
    }
);
desc('Installing assets');
task(
    'deploy-assets',
    function () {
        $result = run('cd {{release_path}} && yarn install --cache-folder .yarn_cache');
        writeln($result);
        $result = run('cd {{release_path}} && chmod 0755 ./node_modules/.bin/encore');
        writeln($result);
        $result = run('cd {{release_path}} && yarn run encore production');
        writeln($result);
    }
);
// endregion

// copy the vendor from the previous release to speed up composer install
set('copy_dirs', ['vendor', 'node_modules']);

// region sf4 app version
set(
    'app_version',
    function () {

        $cmd            = "curl -sS -H 'Authorization: token {{github_token}}' --location --request GET 'https://raw.githubusercontent.com/{{repository_name}}/{{revision}}/config/packages/app.yaml'";
        $configContents = runLocally($cmd);

        return (new ProjectUtils())->parseAppVersion(ProjectUtils::APP_TYPE_SF4, $configContents);
    }
);
// endregion

// region tasks
// set up composer auth
before('deploy:vendors', 'composer-set-auth');

// we need to make sure .env are set correctly in SHARED FILES
before('deploy:symlink', 'database:migrate');
// endregion


recipes/bootstrap_github_action.php
<?php

namespace Deployer;

use Utils\ProjectUtils;
use Deployer\Utility\Httpie;

require realpath(__DIR__ . '/../../vendor/autoload.php');
require realpath(__DIR__ . '/../../vendor/deployer/recipes/autoload.php');
require 'recipe/common.php';

// region environment
$revision          = $_ENV['GITHUB_SHA'] ?? null;
$repositoryName    = $_ENV['GITHUB_REPOSITORY'] ?? null;
$token             = $_ENV['GITHUB_TOKEN'] ?? null;
$slackWebhookToken = $_ENV['SLACK_WEBHOOK_TOKEN'] ?? null;
$githubActor       = $_ENV['GITHUB_ACTOR'] ?? null;
$branch            = $result = preg_replace('%^refs/heads/(.*)%m', '\1', $_ENV['GITHUB_REF'] ?? '');
$start             = microtime(true);

if ($token === null) {
    echo 'Missing GITHUB_TOKEN';
    exit(1);
}
if ($revision === null) {
    echo 'Missing GITHUB_SHA';
    exit(1);
}
if ($repositoryName === null) {
    echo 'Missing GITHUB_REPOSITORY';
    exit(1);
}
if ($branch === '') {
    echo 'Unable to determine branch';
    exit(1);
}

// region determine stage to deploy
$stage = 'staging';
if (in_array($branch, ['production', 'master'])) {
    $stage = 'production';
}
// endregion

// endregion

// set configuration
set('default_stage', $stage);
set('revision', $revision);
set('branch', $branch);
set('repository_name', $repositoryName);
set('user', $githubActor);
set('github_token', $token);
set('slack_webhook_token', $slackWebhookToken);
set('composer_auth_json', $_ENV['COMPOSER_AUTH_JSON'] ?? null);

set(
    'changelog',
    function () {

        $cmd       = "curl -sS -H 'Authorization: token {{github_token}}' --location --request GET 'https://raw.githubusercontent.com/{{repository_name}}/{{revision}}/CHANGELOG.md'";
        $changeLog = runLocally($cmd);

        $projectUtils = new ProjectUtils();

        return $projectUtils->parseChangelog($changeLog);
    }
);

set(
    'app_version_changelog',
    function () {
        $changelog  = get('changelog');
        $appVersion = get('app_version');
        $retValue   = '';

        if ($changelog !== null && $appVersion !== null && array_key_exists($appVersion, $changelog)) {
            $retValue = (new ProjectUtils())->changeLogVersionToString($changelog[$appVersion]);
        }

        return $retValue;
    }
);

set(
    'app_version',
    function () {
        return null;
    }
);

// default was acl - but not enabled by ips containers
set('writable_mode', 'chmod');

// set deploy path
set('deploy_path', '~/deployer');

// clear HOME
// https://github.com/deployphp/deployer/blob/master/src/Support/Unix.php - parseHomeDir uses HOME which is overridden by github
unset($_SERVER['HOME']);

// region tasks

// unlock after failure
after('deploy:failed', 'deploy:unlock');

desc('Update code - note: this overrides default deployer deploy:update_code');
task(
    'deploy:update_code',
    [
        'download-revision',
        'extract-revision',
        'deploy:copy_dirs',
        'create-private-html-symlink',
        'deploy:setup-permissions',
    ]
);

desc('Download revision tar');
task(
    'download-revision',
    function () {

        $repositoryName = get('repository_name', 'xxx');
        $revision       = get('revision', 'xxx');
        $token          = get('github_token', 'xxx');

        run("cd {{release_path}} && curl -sS -H 'Authorization: token $token' --location --request GET 'https://api.github.com/repos/$repositoryName/tarball/$revision' > repo.tar.gz");
    }
);

desc('Extract downloaded revision tar');
task(
    'extract-revision',
    function () {
        $repositoryName = get('repository_name', 'xxx');
        $revision       = get('revision', 'xxx');

        $directionName = str_replace('/', '-', $repositoryName) . '-' . $revision . '/';
        run("cd {{release_path}} && tar -xvf repo.tar.gz  $directionName");
        run("cd {{release_path}} && (mv $directionName{*,.[^.]*,..?*} . 2>/dev/null || true)"); // move all files included dot files (when a project does not have dot files ignore error
        run("cd {{release_path}} && rm -rf $directionName");
        run("cd {{release_path}} && rm repo.tar.gz");
    }
);

desc('Set required symlink for private-html, used for https');
task(
    'create-private-html-symlink',
    function () {
        run("cd {{release_path}} && ln -s public_html private_html");
    }
);

desc('Setting auth for composer for private packages');
task(
    'composer-set-auth',
    function () {
        $composerAuthJson = json_decode(get('composer_auth_json', null), true);
        if ($composerAuthJson !== null && $composerAuthJson !== false) {
            foreach ($composerAuthJson['http-basic'] ?? [] as $host => $config) {
                $username = $config['username'] ?? null;
                $password = $config['password'] ?? null;
                if ($username !== null && $password !== null) {
                    run(sprintf('cd {{release_path}} && composer config http-basic.%1$s %2$s "%3$s"', $host, $username, $password));
                }
            }
        }
    }
);

desc('Setup permissions');
task(
    'deploy:setup-permissions',
    function () {
        // folders
        run("cd {{release_path}} && find . -type d -not \( -path './awstats*' -o -path './.git*' -o -path \"./var/cache*\" \) ! -perm 0755 -exec chmod 755 {} \;");

        // files
        run("cd {{release_path}} && find . -type f -not \( -path './awstats/*' -o -path './logs/*' -o -path './stats/*' -o -path './.git/*' \) ! -perm 0644 -exec chmod 644 {} \;");
    }
);

desc('Create release in github');
task(
    'github:create-release',
    function () use ($start) {
        $stage = get('stage') ?? get('default_stage');
        if ($stage == 'production') {
            writeln('Creating new release <info>{{app_version}}</info>...');

            $result = createRelease(get('github_token'), get('repository_name'), get('app_version'), get('app_version_changelog'));

            if (array_key_exists('html_url', $result)) {
                // region notification on slack
                if (!get('slack_webhook', false)) {
                    return;
                }

                $url  = $result['html_url'];
                $name = $result['name'];

                set('slack_success_text', sprintf('Deployment to *{{target}}* with version `%1$s` successful - see <%2$s|release> - {{duration}}', $name, $url));
            }
            // endregion
        }
    }
)->once();

after('success', 'github:create-release');

set(
    'duration',
    function () use ($start) {
        return sprintf('%1$.2fs', microtime(true) - $start);
    }
);

// @see https://deployer.org/recipes/slack.html
$slackWebhookToken = get('slack_webhook_token', null);
if ($slackWebhookToken !== null) {
    require 'recipe/slack.php';

    set('slack_webhook', $slackWebhookToken);
    set('slack_text', sprintf('_{{user}}_ deploying `{{branch}}` with version `{{app_version}}` to *{{target}}* - _{{stage}}_'));

    after('success', 'slack:notify:success');
    after('deploy:failed', 'slack:notify:failure');

    before('deploy:info', 'slack:notify');

    set('slack_success_text', 'Deployment to *{{target}}* with version `{{app_version}}` successful - {{duration}}');
}

desc('Display additional info');
task(
    'deploy:info-stage',
    function () {
        writeln('Deployment target is <info>{{stage}}</info> deploying app version <info>{{app_version}}</info>');
    }
)->once()
 ->shallow()
 ->setPrivate();
before('deploy:info', 'deploy:info-stage');

// endregion
/**
 * Create release
 *
 * @param string $githubToken
 * @param string $repositoryName
 * @param string $version
 * @param string $changeLogVersionBody
 * @return array
 */
function createRelease(string $githubToken, string $repositoryName, string $version, string $changeLogVersionBody): array
{
    $curl = curl_init();

    $tagName     = ltrim($version, 'v');
    $releaseName = 'v' . $tagName;

    $data = [
        'tag_name'         => $tagName,
        'target_commitish' => 'master',
        'name'             => $releaseName,
        'body'             => $changeLogVersionBody,
        'draft'            => false,
        'prerelease'       => false,
    ];

    curl_setopt_array(
        $curl,
        [
            CURLOPT_URL            => 'https://api.github.com/repos/' . $repositoryName . '/releases',
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING       => "",
            CURLOPT_MAXREDIRS      => 10,
            CURLOPT_TIMEOUT        => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION   => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST  => 'POST',
            CURLOPT_POSTFIELDS     => json_encode($data),
            CURLOPT_HTTPHEADER     => [
                'Content-Type: application/json',
                'User-Agent: deployer',
                'Authorization: Bearer ' . $githubToken,
            ],
        ]
    );

    $response = curl_exec($curl);

    curl_close($curl);

    return json_decode($response, true);
}
